shell/server/installer.js

```diff
@@ -385,6 +388,10 @@ AppInstaller = class AppInstaller {
             (manifest.appTitle && manifest.appTitle.defaultText), manifest.appVersion,
             (manifest.appMarketingVersion && manifest.appMarketingVersion.defaultText));
         }
+
+        if (_this.isPreinstalled) {
```

Comment:
Sorry, I discussed this with @paulproteus in person after he posted this note, but the quick answer is that this doesn't actually update pre-existing users. This function is called when a package has been successfully downloaded to the server, but it does not add UserActions (ie. add it to a user's app list).

To your first question, the code now handles this correctly. The only way the packageId of a preinstalled app is ever changed is either when an admin clicks "Save" on the pre-installed apps admin panel, or when an autoupdate comes over the wire. Local versions are ignored.